drivers/usb/usbport/endpoint.c

```diff
@@ -66,45 +66,171 @@ USBPORT_AllocateBandwidth(IN PDEVICE_OBJECT FdoDevice,
     PUSBPORT_DEVICE_EXTENSION FdoExtension;
     PUSBPORT_ENDPOINT_PROPERTIES EndpointProperties;
     ULONG TransferType;
-    ULONG TotalBusBandwidth;
     ULONG EndpointBandwidth;
+    ULONG TotalBusBandwidth;
+    ULONG MinBandwidth;
+    PULONG Bandwidth;
+    ULONG MaxBandwidth = 0;
+    ULONG ix;
+    ULONG Offset;
+    LONG ScheduleOffset = -1;
     ULONG Period;
+    ULONG Factor;
+    UCHAR Bit;
 
-    DPRINT("USBPORT_AllocateBandwidth: ... \n");
+    DPRINT("USBPORT_AllocateBandwidth: FdoDevice - %p, Endpoint - %p\n",
+           FdoDevice,
+           Endpoint);
 
     FdoExtension = FdoDevice->DeviceExtension;
     EndpointProperties = &Endpoint->EndpointProperties;
     TransferType = EndpointProperties->TransferType;
 
-    if (TransferType == USBPORT_TRANSFER_TYPE_BULK || 
+    if (TransferType == USBPORT_TRANSFER_TYPE_BULK ||
         TransferType == USBPORT_TRANSFER_TYPE_CONTROL ||
-        Endpoint->Flags & ENDPOINT_FLAG_ROOTHUB_EP0)
+        (Endpoint->Flags & ENDPOINT_FLAG_ROOTHUB_EP0) != 0)
     {
         EndpointProperties->ScheduleOffset = 0;
         return TRUE;
     }
 
     TotalBusBandwidth = FdoExtension->TotalBusBandwidth;
     EndpointBandwidth = EndpointProperties->UsbBandwidth;
+
     Period = EndpointProperties->Period;
+    Factor = USB2_FRAMES / Period;
+    ASSERT(Factor);
 
-    DPRINT1("USBPORT_AllocateBandwidth: FIXME. \n");
-    DPRINT1("USBPORT_AllocateBandwidth: Endpoint - %p, Type - %x, TotalBandwidth - %x, EpBandwidth - %x, Period - %x\n",
-           Endpoint,
-           TransferType,
-           TotalBusBandwidth,
-           EndpointBandwidth,
-           Period);
+    for (Offset = 0; Offset < Period; Offset++)
+    {
+        MinBandwidth = TotalBusBandwidth;
+        Bandwidth = &FdoExtension->Bandwidth[Offset * Factor];
+
+        for (ix = 0; *Bandwidth >= EndpointBandwidth; ix++)
+        {
+            if (MinBandwidth > *Bandwidth )
+            {
+                MinBandwidth = *Bandwidth;
+            }
+
+            Bandwidth++;
+
+            if (Factor <= (ix + 1))
+            {
+                if (MinBandwidth > MaxBandwidth )
+                {
+                    MaxBandwidth = MinBandwidth;
+                    ScheduleOffset = Offset;
+
+                    DPRINT("USBPORT_AllocateBandwidth: ScheduleOffset - %X\n",
+                           ScheduleOffset);
+                }
 
-    return TRUE;
+                continue;
```

Comment:
This `continue` has no effect. I wonder whether it's meant for the outer loop, i.e. should be `break`?

Reply:
Confirmed that it should be `break`. Will change.